### PR TITLE
[ECO-511] Add postMessage-based session activation flow

### PR DESCRIPTION
**Pull request by msmolens**
JIRA: [ECO-511](https://mural.atlassian.net/browse/ECO-511)

Add `postMessage`-based session activation flow.

### CODE REVIEW

File: sample/src/app-automation.tsx [hunk 1]
```diff
@@ -111,6 +111,10 @@ class App extends React.Component<{}, AppState> {
     console.log('[handleMessage]', evt);
   };
 
+  handleSessionActivationRequest = async (code: string) => {
+    await createMuralSessionAndNotify(apiClient, this.rpcClient, code);
+  };
```
[the remark this answers is not in the document]
Comment (micaelbergeron) [reply sample/src/app-automation.tsx:116]: Another option would be to have a `<AuthenticatedCanvas />` component that _requires_ the `rpcClient` to be set.

File: packages/mural-canvas/src/lib/session-activation.ts [hunk 2]
```diff
@@ -1,22 +1,33 @@
...
  * Request a session activation context from the MURAL canvas to
  * prepare the canvas to safely activate our session.
  */
 export function muralSessionActivationUrl(
   apiClient: ApiClient,
-  authUrl: URL | string,
+  rpc: boolean,
   muralUrl: URL | string,
+  authUrl?: URL | string,
 ): URL {
```
[the remark this answers is not in the document]
Comment (msmolens) [reply packages/mural-canvas/src/lib/session-activation.ts]: Thanks, I split these into two separate functions.

File: packages/mural-canvas/src/lib/rpc.ts [hunk 3]
```diff
@@ -92,6 +102,8 @@ export default class RpcClient extends EventEmitter {
     this.config.source.addEventListener('message', this.recv);
   };
 
+  notifySessionReady = () => this.rpc('mural.session_activation');
+
```
[the remark this answers is not in the document]
Comment (msmolens) [reply packages/mural-canvas/src/lib/rpc.ts]: I updated this so the `RpcClient` class has only transport-level concerns.

File: packages/mural-canvas/src/lib/rpc.ts [hunk 4]
```diff
@@ -144,8 +156,15 @@ export default class RpcClient extends EventEmitter {
     this.rpcContext = to;
   };
 
-  recv = (evt: MessageEvent<RpcCallback>) => {
+  recv = (evt: MessageEvent<RpcCallback | SessionActivationRequest>) => {
```
[the remark this answers is not in the document]
Comment (msmolens) [reply packages/mural-canvas/src/lib/rpc.ts]: murally now sends an `RpcMessage`. The SDK acknowledges the message by calling `rpc_callback(message)`.

File: sample/src/app-automation.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (msmolens) [reply sample/src/app-automation.tsx]: I moved this handling into the `Canvas` component. When we move this out of WIP I can update the class documentation to describe how to set up each session activation flow.

File: packages/mural-canvas/src/lib/rpc.ts [hunk 5]
```diff
@@ -181,6 +193,19 @@ export default class RpcClient extends EventEmitter {
     this.dispatch();
   };
 
+  rpc_callback(message: RpcMessage, error?: string, context?: RpcContext) {
```
Comment (micaelbergeron) [inline packages/mural-canvas/src/lib/rpc.ts:196]: **nitpick**: I would like the interface to feel happier here, as sending `error: null`, when you want to send `context: {}` seems like a thorn in the side.

What do you think of using an object here as keyword params?
```
rpc_callback(message: RpcMessage, reply?: { error?: string, context?: RpcContext });

# invocations looks quite a bit better
const msg: RpcMessage = …;

rpc_callback(msg); // ACK
rpc_callback(msg, { context: { result: 42 } }); // ACK w/ context
rpc_callback(msg, { error: "I'm not a teapot!" }); // ACK w/ error
```

---

Another option would be to have `rpc_callback_error` to deal with the error case independently.

```
rpc_callback(msg, context);
rpc_callback_error(msg, error, context);
```

File: packages/mural-canvas/src/lib/rpc.ts [hunk 6]
```diff
@@ -181,6 +193,19 @@ export default class RpcClient extends EventEmitter {
...
+    if (!this.config) throw new Error('RpcClient not initialized');
+
+    const callback: RpcCallback = {
+      type: 'mural.rpc_callback',
+      rpcid: message.rpcid,
+      error,
+      context,
```
Comment (micaelbergeron) [inline packages/mural-canvas/src/lib/rpc.ts:203]: **quesiton**: I wonder if we should sent the current `context` as part of the `rpc_callback`.

Comment (msmolens) [reply packages/mural-canvas/src/lib/rpc.ts]: We could do that and perhaps merge in the contents of the `context` parameter.

Alternatively, we could leave control to the client code--they can already explicitly pass the current `context` from `RpcClient`.

File: packages/mural-canvas/src/lib/rpc.ts (hunk 5 above)
Comment (msmolens) [reply packages/mural-canvas/src/lib/rpc.ts]: > an object [...] as keyword params

That's much nicer. I updated this.

File: packages/mural-canvas/src/components/canvas/index.tsx [hunk 7]
```diff
@@ -72,6 +76,24 @@ export class CanvasHost extends React.Component<PropTypes> {
     }
   };
 
+  handleRpcMessage = async (message: RpcMessage) => {
+    const { apiClient, rpcClient } = this.props;
+
+    if (!rpcClient) return;
```
Comment (micaelbergeron) [inline packages/mural-canvas/src/components/canvas/index.tsx:82]: **error**: I think we should `throw new Error('Broken RPC connection, rpcClient missing.');

This is the kind of "WHAT THE HELL" moment that will break behavior anyways, so might as well surface the issue.

File: packages/mural-canvas/src/components/canvas/index.tsx [hunk 8]
```diff
@@ -72,6 +76,24 @@ export class CanvasHost extends React.Component<PropTypes> {
...
+
+    // Handle only session activation request
+    if (message.method !== 'session_activation_request') return;
+
+    const code = message.args?.at(0);
+    if (!code) return;
```
Comment (micaelbergeron) [inline packages/mural-canvas/src/components/canvas/index.tsx:88]: **error**: Let's tell the client this is a incorrect message.

```
return this.rpc_callback(message, { error: "'code' is missing" });
```

File: packages/mural-canvas/src/lib/rpc.ts [hunk 9]
```diff
@@ -181,6 +193,25 @@ export default class RpcClient extends EventEmitter {
     this.dispatch();
   };
 
+  rpc_callback(
```
Comment (micaelbergeron) [inline packages/mural-canvas/src/lib/rpc.ts:196]: **naming**: I'm so sorry, IDK why I said to name this `rpc_callback`, I think we should follow camelCase and call this `rpcCallback`.

Let's fix that before people start using this interface.

It's probably my brain mapping the `mural.rpc_callback` message type.

**Review by micaelbergeron — APPROVED** [review]
LGTM :star:

**Review by farminf — COMMENTED** [review]
looked at it after a while and this is really good work 🎉

**Pull request merged**